designs/standards/errorMessaging.md

```diff
@@ -0,0 +1,91 @@
+
+### Build Errors
+Build errors are reflected in a few locations:
+1. The status bar at the bottom left corner of the IDE
+2. A toast pointing to the status bar
+3. The IDE’s Build panel
+
+![6]
+
+![7]
+
+#### Surfacing Method
+* Create a [`ProgramRunner`](https://upsource.jetbrains.com/idea-ce/file/idea-ce-e97504227f5f68c58cd623c8f317a134b6d440b5/platform/lang-api/src/com/intellij/execution/runners/ProgramRunner.java) and call `setError()`
+
+### Form Errors
+Multiple types of errors can be spawned from JetBrains forms:
+
+* Failed validation
+    * Occurs when field validation is engaged. Our current standard is to run validation when a user attempts to submit a wizard.
+    * Presents an error message and highlights the input box in red.
+    
+![8]
+* Failed data population
+    * Occurs automatically, if the wizard is trying to load data to populate a field and fails
+    * Due to the general error aspect, these errors are also propagated to the Event Log.
+    
+![9]
+* Failed run configuration
+    * Can occur at any point during a run configuration
+    * Presented as a separate error in the bottom of the form. This is not directly tied to a field (like a validation error)
+    
+![10]
+
+#### Surfacing Method
+* Failed validation: return `ValidationInfo` object
+* Failed data population: Current standard is to use a [`ResourceSelector`](https://github.com/aws/aws-toolkit-jetbrains/blob/master/jetbrains-core/src/software/aws/toolkits/jetbrains/ui/ResourceSelector.kt) object. If this fails, this draws a general and validation error (through their respective mechanisms).
+* Failed run configuration: Throw a RuntimeConfigurationError from a class that extends [`LocatableConfigurationBase`](https://upsource.jetbrains.com/idea-ce/file/idea-ce-d00d8b4ae3ed33097972b8a4286b336bf4ffcfab/platform/lang-api/src/com/intellij/execution/configurations/LocatableConfigurationBase.java)
+
+## Guidelines
+
+* Stack traces displayed in the Event Log should not contain any AWS Toolkit- or IDE-specific stack traces as these are not user actionable. These errors should be logged to the `idea.log`.
+* Similarly, the Event Log should display stack traces for any user-actionable errors. These should also be logged to the `idea.log` file.
+    * We should strive to log errors where they're expected. General errors should display in the Event Log, build errors in the build panel, etc.
+    * Alternately, display errors in the event log and have error messages point to the larger stack trace in the Event Log (for instance, in the build panel)
+* We should consistently display errors in expected locations. We may want to modify some existing errors for consistency. For example, the following images show a `sam build` failure (as of August 2019), with the first one from running a SAM application locally vs. building and deploying the SAM app to create a Lambda function:
```

Comment:
Consider tightening this up a little. Example:

* Do display errors in expected locations consistently. 
  * There may be instances in the repo that do not follow convention yet.
  * Good example: (some image(s) here)
  * Bad example:
    * `sam build` failures surface in different portions of the IDE: ![6] ![1]